### DISCUSSION TITLE
Segmentation fault (core dumped) when write png files

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2150] [Q&A]
When I write a png to buffer ,it failed.

The core information :
#0  0x00000000008722fa in putShortMSB ()
#1  0x00000000008726a3 in deflate ()
#2  0x00007f09f9748e4a in png_write_find_filter () at /lib64/libpng15.so.15
#3  0x00007f09f9743fe5 in png_write_row () at /lib64/libpng15.so.15
#4  0x00007f09f97441bf in png_write_rows () at /lib64/libpng15.so.15
#5  0x00007f09fc77b7d3 in write_png_block () at /lib64/libvips.so.42
#6  0x00007f09fc7c9636 in wbuffer_write_thread () at /lib64/libvips.so.42
#7  0x00007f09fc7dad14 in vips_thread_run () at /lib64/libvips.so.42
#8  0x00007f09fc0df540 in g_thread_proxy () at /lib64/libglib-2.0.so.0
#9  0x00007f09fbe57ea5 in start_thread () at /lib64/libpthread.so.0
#10 0x00007f09faaab9fd in clone () at /lib64/libc.so.6

but when I write a jpg, it worked well

I read the png from the cntl.response_attachment and put it into buffer. Then write it back. My code as follow:

`   

    int fd;
    void *start;
    int len = cntl.response_attachment().size();
    fd = open("t.png", O_RDWR);
    struct stat sb;
    cntl.response_attachment().cut_into_file_descriptor(fd, len);
    fstat(fd, &sb);
    start = mmap(NULL, sb.st_size, PROT_READ, MAP_SHARED, fd, 0);
    if(start == MAP_FAILED)
    {
        std::cout << "failed" << std::endl;
    }
    std::cout << start << std::endl;
    close(fd);
    VipsImage *pic = vips_image_new_from_buffer(start, len, "", NULL);
    vips_rotate(pic, &pic, 30, NULL);
    size_t length;
    //vips_image_write_to_buffer(pic, ".jpg", &start, &length, NULL); // this worked
    vips_image_write_to_buffer(pic, ".png", &start, &length, NULL); // this failed
`

**Comment by jcupitt** [comment]
Hello @rebornfz,

It should work -- there is probably an error in your code somewhere.

If you make a complete, runnable example that shows the crash, I can test it.

**Reply by rebornfz** [reply]
Thank for your replay. I have solved the problem.
